src/axom/lumberjack/MPIUtility.cpp

```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
   return charArray;
 }
 
+const char* mpiNonBlockingReceiveMessages(MPI_Comm comm, const int tag)
+{
+  const int mpiTag = (tag == false) ? LJ_TAG : tag;
+  char* charArray = nullptr;
+  int messageSize = -1;
+  MPI_Status mpiStatus;
+
+  // Get size and source of MPI message
+  int mpiFlag = true;
+  MPI_Iprobe(MPI_ANY_SOURCE, tag, comm, &mpiFlag, &mpiStatus);
+
+  if (mpiFlag == true) {
+    MPI_Get_count(&mpiStatus, MPI_CHAR, &messageSize);
+
+    // Setup where to receive the char array
+    charArray = new char[messageSize + 1];
+    charArray[messageSize] = '\0';
+
+    // Receive packed Message
+    MPI_Recv(charArray,
+            messageSize,
+            MPI_CHAR,
+            mpiStatus.MPI_SOURCE,
+            mpiTag,
+            comm,
+            &mpiStatus);
```

Comment:
As I understand the MPI API, this is actually a blocking `MPI_Recv` [call](https://www.mpich.org/static/docs/v3.4/www3/MPI_Recv.html)? So this `mpiNonBlockingReceiveMessages` function is currently blocking to receive messages.

Reply:
Yes, that's correct.  The non-blocking part is the call to MPI_Iprobe, but then the Recv is blocking.  My intent here is to be sure that the receive is fully finished before anything else is done, but to not block any further execution if there are no messages to be received (i.e. when mpiFlag is false).  I can change the function name to clarify the intent here.

Reply:
To clarify the above point, MPI_Iprobe is used instead of MPI_probe because the former will return with an mpiFlag value regardless of whether messages need to be received, whereas the latter is a blocking call that will only return if there is a message to be received.

Reply:
This does also make me think about renaming this communicator to something like "NonCollectiveCommunicator" rather than "NonBlockingCommunicator".  It's true that it calls these non-blocking functions, but I think the main feature is actually that we don't rely on collective calls to communicate messages to root.

Reply:
I like that idea.